sqlparser/lib/src/analysis/types/resolving_visitor.dart

```diff
@@ -131,12 +131,28 @@ class TypeResolver extends RecursiveVisitor<TypeExpectation, void> {
   }
 
   @override
-  void visitSetComponent(SetComponent e, TypeExpectation arg) {
+  void visitSingleColumnSetComponent(
+      SingleColumnSetComponent e, TypeExpectation arg) {
     visit(e.column, const NoTypeExpectation());
     _lazyCopy(e.expression, e.column);
     visit(e.expression, const NoTypeExpectation());
   }
 
+  @override
+  void visitMultiColumnSetComponent(
+      MultiColumnSetComponent e, TypeExpectation arg) {
+    visitList(e.columns, const NoTypeExpectation());
+    if (e.rowValue is Tuple) {
+      final expressions = (e.rowValue as Tuple).expressions;
+      for (final (idx, expression) in expressions.indexed) {
+        _lazyCopy(expression, e.columns.elementAtOrNull(idx));
+      }
+    } else if (e.rowValue is SubQuery) {
+      // TODO: handle subquery case
```

Comment:
I think this can be handled similarly to what we're doing for inserts (see `visitInsertStatement`). Basically, visit the columns first which should give us their type since they're direct references to a schema we already know. Then, visit the RHS with a `SelectTypeExpectation` which is deconstructed in `visitTuple` and `visitSelectStatement`, so we can get away with a single visit and don't need to distinguish the two cases here at all.

Reply:
Thank you. I have mostly copied the code from `visitInsertStatement` now and all tests pass :).
It would be nevertheless helpful if you could look into the current implementation of `TypeResolver.visitMultiColumnSetComponent`. I am especially unsure about the calls to `_handleColumn` , as the tests pass with and without it.

Reply:
> I am especially unsure about the calls to `_handleColumn` , as the tests pass with and without it.

I think it doesn't make a difference here since we'll always have references to schema columns with a fixed type. For some other column references like those to a CTE or a compound select, calling `_handleColumn` is necessary to resolve their types. It also closes the gap between the AST model and the graph of column references.

So it doesn't hurt to have it.

Reply:
Alright, thanks for your explanation.